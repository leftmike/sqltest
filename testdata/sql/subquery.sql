--
-- Test subqueries
--
-- {{if eq Dialect "sqlite3"}}{{Skip}}{{end}}

DROP TABLE IF EXISTS tbl1;

CREATE TABLE tbl1 (c1 int PRIMARY KEY, c2 int, c3 int, c4 int);

INSERT INTO tbl1 VALUES
    (1, 10, 11, 100),
    (2, 20, 22, 200),
    (3, 30, 33, 300),
    (4, 40, 44, 400),
    (5, 50, 55, 500),
    (6, 60, 66, 600),
    (7, 70, 77, 700),
    (8, 80, 88, 800),
    (9, 90, 99, 900);

SELECT * FROM tbl1;

DROP TABLE IF EXISTS tbl2;

CREATE TABLE tbl2 (c1 int PRIMARY KEY, c2 TEXT);

INSERT INTO tbl2 VALUES
    (10, 'ten'),
    (20, 'twenty'),
    (30, 'thirty'),
    (40, 'forty'),
    (50, 'fifty');

SELECT * FROM tbl2;

SELECT col2, col1 FROM (SELECT c4, c2 FROM tbl1 WHERE c1 >= 4) AS t1 (col1, col2);

-- {{Fail .Test}}
SELECT col2, col1, c2 FROM (SELECT c4, c2 FROM tbl1) AS t1 (col1, col2);

-- {{Fail .Test}}
SELECT col2, col1 FROM (SELECT c4, c2 FROM tbl1 WHERE c1 >= 4) AS t1 (col1, col2) WHERE c2 <= 70;

SELECT col2, col1 FROM (SELECT c4, c2 FROM tbl1 WHERE c1 >= 4) AS t1 (col1, col2) WHERE col2 <= 70;

SELECT col2, col1
    FROM (SELECT c4 + c3, tbl2.c2 FROM tbl1 JOIN tbl2 ON tbl1.c2 = tbl2.c1) AS t1 (col1, col2)
    WHERE col1 % 2 = 1;

SELECT * FROM (VALUES (1, 2, 3), (4, 5, 6), (7, 8, 9)) AS t1 (c1, c2, c3);

SELECT *
    FROM (SELECT c3, c2 + c1 FROM (VALUES (1, 2, 3), (4, 5, 6), (7, 8, 9)) AS t1 (c1, c2, c3))
    AS t2 (col1, col2);

SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE c1 = tbl1.c2) AS e3 FROM tbl1;

SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE c1 = 10) AS e3 FROM tbl1;

SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE c1 = 60) AS e3 FROM tbl1;

SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE tbl1.c2 > 50) AS e3 FROM tbl1;

SELECT c1, c2 FROM tbl1 WHERE c2 IN (SELECT c1 FROM tbl2);

SELECT c1, c2 FROM tbl1 WHERE c2 NOT IN (SELECT c1 FROM tbl2);

SELECT c1, c2 IN (SELECT c1 FROM tbl2) AS e1 FROM tbl1;

SELECT c1, c2 NOT IN (SELECT c1 FROM tbl2) AS e1 FROM tbl1;

DROP TABLE IF EXISTS tbl3;

CREATE TABLE tbl3 (c1 int PRIMARY KEY, c2 TEXT, c3 int);

INSERT INTO tbl3 VALUES
    (10, 'ten', 11),
    (20, NULL, 22),
    (30, 'thirty', 33),
    (40, NULL, NULL),
    (50, 'fifty', 55);

SELECT * FROM tbl3;

SELECT c1, c3 FROM tbl1 WHERE c3 IN (SELECT c3 FROM tbl3);

SELECT c1, c3 FROM tbl1 WHERE c3 NOT IN (SELECT c3 FROM tbl3);

SELECT c1, c3 IN (SELECT c3 FROM tbl3) AS e1 FROM tbl1;

SELECT c1, c3 NOT IN (SELECT c3 FROM tbl3) AS e1 FROM tbl1;

-- ANY

SELECT c1, c2 FROM tbl1 WHERE c2 >= ANY (SELECT c1 + 20 FROM tbl2);

SELECT c1, c2 < ANY (SELECT c1 - 10 FROM tbl2) AS e1 FROM tbl1;

-- ALL

SELECT c1, c2 FROM tbl1 WHERE c2 >= ALL (SELECT c1 + 20 FROM tbl2);

SELECT c1, c2 < ALL (SELECT c1 + 40 FROM tbl2) AS e1 FROM tbl1;
