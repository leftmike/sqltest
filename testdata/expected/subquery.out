--
-- Test subqueries
--
-- {{if eq Dialect "sqlite3"}}{{Skip}}{{end}}
DROP TABLE IF EXISTS tbl1;
CREATE TABLE tbl1 (c1 int PRIMARY KEY, c2 int, c3 int, c4 int);
INSERT INTO tbl1 VALUES
    (1, 10, 11, 100),
    (2, 20, 22, 200),
    (3, 30, 33, 300),
    (4, 40, 44, 400),
    (5, 50, 55, 500),
    (6, 60, 66, 600),
    (7, 70, 77, 700),
    (8, 80, 88, 800),
    (9, 90, 99, 900);
SELECT * FROM tbl1;
   c1 c2 c3  c4
   -- -- --  --
 1  1 10 11 100
 2  2 20 22 200
 3  3 30 33 300
 4  4 40 44 400
 5  5 50 55 500
 6  6 60 66 600
 7  7 70 77 700
 8  8 80 88 800
 9  9 90 99 900
(9 rows)
DROP TABLE IF EXISTS tbl2;
CREATE TABLE tbl2 (c1 int PRIMARY KEY, c2 TEXT);
INSERT INTO tbl2 VALUES
    (10, 'ten'),
    (20, 'twenty'),
    (30, 'thirty'),
    (40, 'forty'),
    (50, 'fifty');
SELECT * FROM tbl2;
   c1     c2
   --     --
 1 10    ten
 2 20 twenty
 3 30 thirty
 4 40  forty
 5 50  fifty
(5 rows)
SELECT col2, col1 FROM (SELECT c4, c2 FROM tbl1 WHERE c1 >= 4) AS t1 (col1, col2);
   col2 col1
   ---- ----
 1   40  400
 2   50  500
 3   60  600
 4   70  700
 5   80  800
 6   90  900
(6 rows)
-- {{Fail .Test}}
SELECT col2, col1, c2 FROM (SELECT c4, c2 FROM tbl1) AS t1 (col1, col2);
-- {{Fail .Test}}
SELECT col2, col1 FROM (SELECT c4, c2 FROM tbl1 WHERE c1 >= 4) AS t1 (col1, col2) WHERE c2 <= 70;
SELECT col2, col1 FROM (SELECT c4, c2 FROM tbl1 WHERE c1 >= 4) AS t1 (col1, col2) WHERE col2 <= 70;
   col2 col1
   ---- ----
 1   40  400
 2   50  500
 3   60  600
 4   70  700
(4 rows)
SELECT col2, col1
    FROM (SELECT c4 + c3, tbl2.c2 FROM tbl1 JOIN tbl2 ON tbl1.c2 = tbl2.c1) AS t1 (col1, col2)
    WHERE col1 % 2 = 1;
     col2 col1
     ---- ----
 1  fifty  555
 2    ten  111
 3 thirty  333
(3 rows)
SELECT * FROM (VALUES (1, 2, 3), (4, 5, 6), (7, 8, 9)) AS t1 (c1, c2, c3);
   c1 c2 c3
   -- -- --
 1  1  2  3
 2  4  5  6
 3  7  8  9
(3 rows)
SELECT *
    FROM (SELECT c3, c2 + c1 FROM (VALUES (1, 2, 3), (4, 5, 6), (7, 8, 9)) AS t1 (c1, c2, c3))
    AS t2 (col1, col2);
   col1 col2
   ---- ----
 1    3    3
 2    6    9
 3    9   15
(3 rows)
SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE c1 = tbl1.c2) AS e3 FROM tbl1;
   c1 c2    e3
   -- --    --
 1  1 10  true
 2  2 20  true
 3  3 30  true
 4  4 40  true
 5  5 50  true
 6  6 60 false
 7  7 70 false
 8  8 80 false
 9  9 90 false
(9 rows)
SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE c1 = 10) AS e3 FROM tbl1;
   c1 c2   e3
   -- --   --
 1  1 10 true
 2  2 20 true
 3  3 30 true
 4  4 40 true
 5  5 50 true
 6  6 60 true
 7  7 70 true
 8  8 80 true
 9  9 90 true
(9 rows)
SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE c1 = 60) AS e3 FROM tbl1;
   c1 c2    e3
   -- --    --
 1  1 10 false
 2  2 20 false
 3  3 30 false
 4  4 40 false
 5  5 50 false
 6  6 60 false
 7  7 70 false
 8  8 80 false
 9  9 90 false
(9 rows)
SELECT c1, c2, EXISTS(SELECT 1 FROM tbl2 WHERE tbl1.c2 > 50) AS e3 FROM tbl1;
   c1 c2    e3
   -- --    --
 1  1 10 false
 2  2 20 false
 3  3 30 false
 4  4 40 false
 5  5 50 false
 6  6 60  true
 7  7 70  true
 8  8 80  true
 9  9 90  true
(9 rows)
SELECT c1, c2 FROM tbl1 WHERE c2 IN (SELECT c1 FROM tbl2);
   c1 c2
   -- --
 1  1 10
 2  2 20
 3  3 30
 4  4 40
 5  5 50
(5 rows)
SELECT c1, c2 FROM tbl1 WHERE c2 NOT IN (SELECT c1 FROM tbl2);
   c1 c2
   -- --
 1  6 60
 2  7 70
 3  8 80
 4  9 90
(4 rows)
SELECT c1, c2 IN (SELECT c1 FROM tbl2) AS e1 FROM tbl1;
   c1    e1
   --    --
 1  1  true
 2  2  true
 3  3  true
 4  4  true
 5  5  true
 6  6 false
 7  7 false
 8  8 false
 9  9 false
(9 rows)
SELECT c1, c2 NOT IN (SELECT c1 FROM tbl2) AS e1 FROM tbl1;
   c1    e1
   --    --
 1  1 false
 2  2 false
 3  3 false
 4  4 false
 5  5 false
 6  6  true
 7  7  true
 8  8  true
 9  9  true
(9 rows)
DROP TABLE IF EXISTS tbl3;
CREATE TABLE tbl3 (c1 int PRIMARY KEY, c2 TEXT, c3 int);
INSERT INTO tbl3 VALUES
    (10, 'ten', 11),
    (20, NULL, 22),
    (30, 'thirty', 33),
    (40, NULL, NULL),
    (50, 'fifty', 55);
SELECT * FROM tbl3;
   c1     c2 c3
   --     -- --
 1 10    ten 11
 2 20        22
 3 30 thirty 33
 4 40          
 5 50  fifty 55
(5 rows)
SELECT c1, c3 FROM tbl1 WHERE c3 IN (SELECT c3 FROM tbl3);
   c1 c3
   -- --
 1  1 11
 2  2 22
 3  3 33
 4  5 55
(4 rows)
SELECT c1, c3 FROM tbl1 WHERE c3 NOT IN (SELECT c3 FROM tbl3);
  c1 c3
  -- --
(no rows)
SELECT c1, c3 IN (SELECT c3 FROM tbl3) AS e1 FROM tbl1;
   c1   e1
   --   --
 1  1 true
 2  2 true
 3  3 true
 4  4     
 5  5 true
 6  6     
 7  7     
 8  8     
 9  9     
(9 rows)
SELECT c1, c3 NOT IN (SELECT c3 FROM tbl3) AS e1 FROM tbl1;
   c1    e1
   --    --
 1  1 false
 2  2 false
 3  3 false
 4  4      
 5  5 false
 6  6      
 7  7      
 8  8      
 9  9      
(9 rows)
-- ANY
SELECT c1, c2 FROM tbl1 WHERE c2 >= ANY (SELECT c1 + 20 FROM tbl2);
   c1 c2
   -- --
 1  3 30
 2  4 40
 3  5 50
 4  6 60
 5  7 70
 6  8 80
 7  9 90
(7 rows)
SELECT c1, c2 < ANY (SELECT c1 - 10 FROM tbl2) AS e1 FROM tbl1;
   c1    e1
   --    --
 1  1  true
 2  2  true
 3  3  true
 4  4 false
 5  5 false
 6  6 false
 7  7 false
 8  8 false
 9  9 false
(9 rows)
-- ALL
SELECT c1, c2 FROM tbl1 WHERE c2 >= ALL (SELECT c1 + 20 FROM tbl2);
   c1 c2
   -- --
 1  7 70
 2  8 80
 3  9 90
(3 rows)
SELECT c1, c2 < ALL (SELECT c1 + 40 FROM tbl2) AS e1 FROM tbl1;
   c1    e1
   --    --
 1  1  true
 2  2  true
 3  3  true
 4  4  true
 5  5 false
 6  6 false
 7  7 false
 8  8 false
 9  9 false
(9 rows)
